### PR TITLE
Features/vault enterprise

### LINKED ISSUE #58585: Vault Enterprise Integration [FEATURE REQUEST]
**Is your feature request related to a problem? Please describe.**
In order to communicate with Vault Enterprise namespace,  a namespace must be declared as part of the headers, as X-Vault-Namespace. Info: https://www.vaultproject.io/docs/enterprise/namespaces

**Describe the solution you'd like**
The files runners/vault.py and utils/vault.py need to be updated so that the requests to vault pass these headers in where necessary.

**Describe alternatives you've considered**
None. If someone wants to use Salt and Vault Enterprise, this change would need to be implemented.

**Additional context**
A PR has been submitted to address this issue: https://github.com/saltstack/salt/pull/58586

I hope I'm doing this right.

### PR DESCRIPTION
**Pull request by eadderley-tc**
### What does this PR do?
This PR enables Salt to communicate with an Enterprise Vault via X-Vault-Namespace headers.

### What issues does this PR fix or reference?
Fixes: [58585](https://github.com/saltstack/salt/issues/58585)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

Not entirely sure how to write tests for this, as the headers don't break anything if no namespace is supplied and Vault doesn't complain if namespaces aren't enabled.

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/utils/vault.py [hunk 1]
```diff
@@ -310,6 +322,17 @@ def make_request(
...
     url = "{0}/{1}".format(vault_url, resource)
     headers = {"X-Vault-Token": str(token), "Content-Type": "application/json"}
+    # Vault Enterprise requires a namespace
+    if "namespace" in args:
+        namespace = args["namespace"]
+    else:
+        try:
+            namespace = __opts__.get("vault").get("namespace", None)
+        except (TypeError, AttributeError):
+            # Don't worry about setting verify if it doesn't exist
```
Comment (mchugh19) [inline salt/utils/vault.py:332]: Nitpick: Comment doesn't appear to match code (or be needed)

File: salt/utils/vault.py [hunk 2]
```diff
@@ -310,6 +322,17 @@ def make_request(
...
+            pass 
+    if namespace is not None:
```
Comment (DmitryKuzmenko) [inline salt/utils/vault.py:334]: You'll get `"namespace"` is not defined exception here iv `"vault"` not in `__opts__`.

File: salt/utils/vault.py [hunk 3]
```diff
@@ -310,6 +322,17 @@ def make_request(
...
+        headers["X-Vault-Namespace"] = namespace
```
Comment (DmitryKuzmenko) [inline salt/utils/vault.py:335]:
```suggestion
    # Vault Enterprise requires a namespace
    try:
        headers["X-Vault-Namespace"] = args.get("namespace") or __opts__["vault"]["namespace"]
    except (TypeError, AttributeError):
        # Don't worry about setting namespace if it doesn't exist
```

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]

**Comment by DmitryKuzmenko** [comment]
@eadderley-tc thank you for PR. I've put some comments to my review.
Also about tests: in this case it's needed to add or update the existing tests to cover your changes. I.e. check that the code works as expected in all the cases you've added: there's no namespace specified, there's a namespace specified, exception cases.

**Comment by eadderley-tc** [comment]
@DmitryKuzmenko We've added some tests and cleaned up the linting. Sorry about the delay, got pulled off for project work and illness. Can you explain what's going wrong in that ci/pre-commit?

**Comment by sagetherage** [comment]
@dwoz or @s0undt3ch can you review this PR please? @eadderley-tc Dmitry is unfortunately no longer on the project, but looking to get this reviewed. Pre-commit can be pain, if it is helpful please attend or get with @waynew on IRC, Community Slack or at the Test Clinics we hold on Tue/Thu via Twitch.

File: salt/utils/vault.py (hunk 1 above)
Comment (s0undt3ch) [reply salt/utils/vault.py]: Yes, please adjust or remove the comment.

File: salt/utils/vault.py [hunk 4]
```diff
@@ -151,6 +162,7 @@ def _use_local_config():
                     ]
             return {
                 "url": __opts__["vault"]["url"],
+                "namespace": __opts__["vault"]["namespace"],
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply salt/utils/vault.py]: This is a valid point that needs to be addressed

**Comment by s0undt3ch** [comment]
Please remove the changes under `doc/man` those get generated automatically.

And there's also conflicts that need to be solved.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Please remove the changes under doc/man those get generated automatically.

And there's also conflicts that need to be solved.

**Comment by eadderley-tc** [comment]
> Please remove the changes under doc/man those get generated automatically.

What do they get automatically generated from?

**Comment by vveliev-tc** [comment]
@s0undt3ch  not sure why some tests are failing, what is currently outstanding for this PR?

**Comment by vveliev-tc** [comment]
re-run centos7-py3-pycryptodome-pytest

**Comment by s0undt3ch** [comment]
Can you please rebase against the latest master branch changes? Thanks!

**Comment by vveliev-tc** [comment]
@s0undt3ch, are we good to merge or need to rebase again?

File: salt/runners/vault.py [hunk 5]
```diff
@@ -52,6 +52,8 @@ def generate_token(
     try:
         config = __opts__.get("vault", {})
         verify = config.get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = config.get("namespace", None)
```
Comment (s0undt3ch) [inline salt/runners/vault.py:56]:
```suggestion
        namespace = config.get("namespace")
```
That extra `, None` is not required. The above code does exactly the same

File: salt/utils/vault.py [hunk 6]
```diff
@@ -103,6 +103,7 @@ def _get_token_and_url_from_master():
         "url": result["url"],
         "token": result["token"],
         "verify": result.get("verify", None),
+        "namespace": result.get("namespace", None),
```
Comment (s0undt3ch) [inline salt/utils/vault.py:106]:
```suggestion
        "namespace": result.get("namespace"),
```

File: salt/utils/vault.py [hunk 7]
```diff
@@ -117,6 +118,8 @@ def get_vault_connection():
 
     def _use_local_config():
         log.debug("Using Vault connection details from local config")
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
Comment (s0undt3ch) [inline salt/utils/vault.py:122]:
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

File: salt/utils/vault.py [hunk 8]
```diff
@@ -297,6 +310,7 @@ def make_request(
         connection = get_cache()
     token = connection["token"] if not token else token
     vault_url = connection["url"] if not vault_url else vault_url
+    namespace = connection["namespace"] if not namespace else namespace
```
Comment (s0undt3ch) [inline salt/utils/vault.py:313]:
```suggestion
    namespace = namespace or connection["namespace"]
```

File: salt/utils/vault.py [hunk 9]
```diff
@@ -360,10 +376,14 @@ def _selftoken_expired():
     """
     try:
         verify = __opts__["vault"].get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
Comment (s0undt3ch) [inline salt/utils/vault.py:380]:
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

File: salt/utils/vault.py [hunk 10]
```diff
@@ -380,10 +400,14 @@ def _wrapped_token_valid():
...
```
Comment (s0undt3ch) [inline salt/utils/vault.py:404]:
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

File: tests/integration/files/file/base/Issues_55775.txt.bak [hunk 11]
```diff
@@ -0,0 +1 @@
+foo
```
Comment (s0undt3ch) [inline tests/integration/files/file/base/Issues_55775.txt.bak:1]: Why is this file here?

File: tests/unit/runners/test_vault.py [hunk 12]
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
...
+        """
+        mock = _mock_json_response(
+            {"auth": {"client_token": "test", "renewable": False, "lease_duration": 0}}
+        )
+        supplied_config = {"namespace": "test_namespace"}
+        with patch("requests.post", mock):
+            with patch.dict(vault.__opts__["vault"], supplied_config):
+                result = vault.generate_token("test-minion", "signature")
+                log.debug("generate_token result: %s", result)
+                self.assertTrue(isinstance(result, dict))
```
Comment (s0undt3ch) [inline tests/unit/runners/test_vault.py:302]:
```suggestion
                self.assertIsInstance(result, dict)
```

File: tests/unit/runners/test_vault.py [hunk 13]
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
...
+                self.assertFalse("error" in result)
```
Comment (s0undt3ch) [inline tests/unit/runners/test_vault.py:303]:
```suggestion
                self.assertNotIn("error", result)
```

File: tests/unit/runners/test_vault.py [hunk 14]
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
...
+                self.assertTrue("token" in result)
```
Comment (s0undt3ch) [inline tests/unit/runners/test_vault.py:304]:
```suggestion
                self.assertIn("token", result)
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Sorry. Minor nitpicks(which actually improve readability)

File: salt/runners/vault.py (hunk 5 above)
Comment (eadderley-tc) [reply salt/runners/vault.py:56]: Neat. I'm pretty sure I added that for some of the testing, but if that's mechanically true, good to know.

Comment (s0undt3ch) [reply salt/runners/vault.py:56]: yep, it is:
```python
>>> a = {'b': 1}
>>> a.get('b')
1
>>> a.get('c')
>>> a.get('c', None)
>>>
```

File: tests/integration/files/file/base/Issues_55775.txt.bak (hunk 11 above)
Comment (eadderley-tc) [reply tests/integration/files/file/base/Issues_55775.txt.bak]: What file? :angel:

**Comment by eadderley-tc** [comment]
@s0undt3ch How's this look?

File: salt/runners/vault.py [hunk 15]
```diff
@@ -267,10 +277,15 @@ def _selftoken_expired():
     """
     try:
         verify = __opts__["vault"].get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
Comment (s0undt3ch) [inline salt/runners/vault.py:281]:
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Final change and we're good

**Comment by vveliev-tc** [comment]
re-run ubuntu1604-py3-tcp-pytest

**Comment by vveliev-tc** [comment]
re-run windows2016-py3-pytest

**Comment by vveliev-tc** [comment]
re-run windows2019-py3-pytest

**Comment by vveliev-tc** [comment]
s0undt3ch - is there anything we should do about failing tests?

**Comment by eadderley-tc** [comment]
@s0undt3ch As far as we can tell (asked on the Slack too) the remaining failing tests are broken. Can we get another review?

**Review by s0undt3ch — DISMISSED** [review]

**Comment by eadderley-tc** [comment]
@s0undt3ch Latest master commits merged in

File: salt/modules/vault.py [hunk 16]
```diff
@@ -49,6 +50,12 @@
 
         .. versionadded:: 2018.3.0
 
+    namespaces
+        Optional Vault Namespace. Used with Vault enterprice
+
+        For detail please see:
+        https://www.vaultproject.io/docs/enterprise/namespaces
```
Comment (Ch3LL) [inline salt/modules/vault.py:57]:
```suggestion
        https://www.vaultproject.io/docs/enterprise/namespaces
        
        .. versionadded:: 3004
```

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]
Looks like there is a just one last lint failure, probably caused by my suggestion. Apologies on that.

**Review by Ch3LL — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]
